---
output:
  html_document:
    highlight: haddock
    toc: true
    toc_depth: 6
    toc_float: true
    theme: cerulean

params:
  v: NA
---
```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE,warning=FALSE,message=FALSE,results='asis',fig.align="center")
knitr::opts_chunk$set(fig.width=9, fig.height=6)
library(shiny)
library(grid)
library(gridExtra)
library(glue)
library(DT)
#library(pander)
#panderOptions('knitr.auto.asis', FALSE)
library(factoextra)
library(pheatmap)
library(ggplot2)
library(clusterProfiler)
```


```{r}
#part 1


if(!("havePart1" %in% names(params$v))){
  cat("#### There is no result or wrong in the data.")
  cat("\n\n")
}else{
  
  cat(glue::glue("# The DGE anlysis report of {params$v$geoid}"), "\n")
  cat("\n\n")
  
  cat("## 1. Data download  \n\n")
  
  cat(glue::glue("### 1.1 {params$v$geoid}"), "\n")
  cat("\n\n")
  
  a<-ncol(params$v$dataExprN)
  cat(glue::glue(" {params$v$geoid} Series Matrix File is downloaded from NCBI GEO database, including {a} GEO Samples."), "\n")
  cat("\n\n")
  
  cat("Here is a part of the expression matrix. You can download the full table of expression data from MephasGEO application. \n\n")
  
  if(ncol(params$v$dataExprN)>=10){
    print(knitr::kable(head(params$v$dataExprN)[,1:10]))
  }else{
    print(knitr::kable(head(params$v$dataExprN)))
  }
  
  cat("\n\n")
  
  
  cat("### 1.2 Boxplot of expression data  \n\n")
  cat("The probe ID of gene microarray is converted into the corresponding gene ID (gene symbol), ")
  
  if(params$v$log2 == TRUE){
    cat("and the data is transformed into the binary logarithm. \n\n")
  }else{
    cat("and there is no need to transform data into the binary logarithm. \n\n")
    
  }
  
  
  cat("Here is the plot drawn by expression data **before group normalization**. \n\n")
  boxplot(params$v$dataExpr,outline=FALSE, notch=T, las=2)
  cat("\n\n")
  
  cat("Here is the plot drawn by expression data **after group normalization**. \n\n")
  boxplot(params$v$dataExprN,outline=FALSE, notch=T, las=2)
  cat("\n\n")
  
  
}


```


```{r fig.width=12, fig.height=5}
#part2 grouping

if("havePart2" %in% names(params$v)){
  cat("## 2.Sample Grouping and clustering \n\n")
  
  cat("### 2.1 Sample Grouping and visualization \n\n")
  
  a<-paste0(params$v$selectfeature,collapse = "/")
  
  cat(glue::glue("The feature \"{a}\" was selected for sample grouping. "), "\n")
  cat("\n\n")
  
  
  aa<-as.list(summary(params$v$group.list))
  
  for (ii in seq_len(length(levels(params$v$group.list)))) {
    cat(glue::glue("There are {aa[ii]} samples in the group of \"{names(aa[ii])}\". "))#, "\n")
  }
  cat("\n\n")
  
  if(params$v$diffchoice == "choiceB"){
    
  if(!is.null(params$v$pair.list)){
  cat(glue::glue("The batch effect selected was \"{params$v$select_batch_name}\". "))#, "\n")
  #cat("\n\n")
  }
    
  }
  
  if(!is.null(params$v$group.rand)){
  cat(glue::glue("The random effect selected was \"{params$v$select_rand_name}\". "), "\n")
  cat("\n\n")
  }
  
  
  cat("### 2.2 Clustering and quality control \n\n")
  
  
  
  
  cat("#### Hierarchical Clustering \n\n")
  par(mar=c(5,5,5,50))
  a<-fviz_dend(params$v$hc,k =length(levels(params$v$group.list)),
                #cex = 0.5, 
                #          k_colors = c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3",
                #            "#FF7F00", "#FFFF33", "#A65628", "#F781BF", "#999999"),
                color_labels_by_k = TRUE, 
                repel = TRUE,
                horiz = TRUE,
                rect = TRUE)
  print(a)
  cat("\n\n")
  cat("#### Principal component analysis (PCA) \n\n")
  par()
  a<-fviz_pca_ind(params$v$dataPCA,
                 col.ind = params$v$group.list, # 颜色对应group信息（这里不对。应该调整顺序
                 #palette = c("#00AFBB",  "#FC4E07"),
                 addEllipses = TRUE, # Concentration ellipses
                 ellipse.type = "confidence",
                 legend.title = "Group",## Legend名称
                 repel = TRUE
    )
   print(a)
   cat("\n\n")
   cat("Tips: Your samples in different groups should be devided to their groups in these plots. If not, it indicated that the result of the DGE analysis may be not so good.")
   cat("\n\n")
}



```


```{r}
#part 3 DGE analysis result

if("havePart3" %in% names(params$v)){
  cat("## 3.The DGE analysis using Limma \n\n")
  cat(glue::glue("Limma was used to do the DGE analysis with the contrast groups \"{params$v$select_contrast_name}\".The p-value was {params$v$pvalue} and the LogFC was 1 in defalt. "), "\n")
  cat("\n\n")
  
  cat("### 3.1 The result \n\n")
  cat(glue::glue("As the result, there are {nrow(params$v$diffSig)} differential expression genes, with {nrow(params$v$diffUp)} upregulation genes and {nrow(params$v$diffDown)} downregulation genes"), "\n")
cat("\n\n")
  
  
  cat("Here is a part of the result of upregulation genes.")
  print(knitr::kable(head(params$v$diffUp)))
  cat("\n\n")
  
  cat("Here is a part of the result of downregulation genes.")
  print(knitr::kable(head(params$v$diffDown)))
  cat("\n\n")
  
  cat("You can download the full table of result from MephasGEO application. \n\n")
  
  
  cat("### 3.2 Volcano plot \n\n")
  
   a<-ggplot(data= params$v$diff_result, aes(x = -log10(adj.P.Val), y = logFC, color = change)) +
        geom_point(alpha=0.8, size = 1) +
        theme_bw(base_size = 15) +
        theme(plot.title=element_text(hjust=0.5),   #  标题居中
              panel.grid.minor = element_blank(),
              panel.grid.major = element_blank()) + # 网格线设置为空白
        geom_hline(yintercept= 0 ,linetype= 2 ) +
        scale_color_manual(name = "", 
                           values = c("red", "green", "black"),
						   limits = c("UP", "DOWN", "NOT")) +
		xlim(0,params$v$xMax) + 
        ylim(-params$v$yMax,params$v$yMax) +
        labs(title = 'Volcano', x = '-Log10(adj.P.Val)', y = 'LogFC')
    print(a)
    cat("\n\n")
    cat("Tips: Each dot in the graph represents a gene. Red dots represent genes that are significantly up-regulated in expression. Green dots represent genes that are significantly down-regulated in expression. Black dots represent genes that have no statistically significant differences.")
    cat("\n\n")
    
  cat("### 3.3 Heatmap \n\n")
  
  par()
  a<-pheatmap(params$v$index_matrix,
          show_colnames =F,
          show_rownames = F,
          cluster_cols = T, 
          annotation_col=params$v$anno)
  print(a)
  cat("\n\n")
  cat("Tips: The column in the heat map represents each sample. The row represents the selected differential gene. The color represents the quantity of gene expression: red represents the high expression of the differential gene in the samples, and blue represents the low expression of the differential gene in the samples.")

  
}



```


```{r fig.width=12, fig.height=5}
#part 4 

if("havePart4" %in% names(params$v)){
  cat("## 4. Enrichment analysis \n\n")
  
  
  cat("### 4.1 GO enrichment analysis \n\n")
  if(!(nrow(params$v$gotab_up)==0)){
    cat("#### GO enriched by Upregular genes \n\n")
    cat("Here is a part of the result.\n\n")
    print(knitr::kable(head(params$v$gotab_up)))
    cat("\n\n")
    
    par()
    a<-barplot(params$v$go_up, split="ONTOLOGY") +facet_grid(ONTOLOGY~., scale="free")
    print(a)
    cat("\n\n")
    testup<-FALSE
  }else{
    testup<-TRUE
  }
  
  if(!(nrow(params$v$gotab_down)==0)){
    cat("#### GO enriched by Downregular genes \n\n")
    cat("Here is a part of the result.\n\n")
    print(knitr::kable(head(params$v$gotab_down)))
    cat("\n\n")
    
    par()
    a<-barplot(params$v$go_down, split="ONTOLOGY") +facet_grid(ONTOLOGY~., scale="free")
    print(a)
    cat("\n\n")
    testdown<-FALSE
  }else{
    testdown<-TRUE
  }
  
  
  if(testup && testdown){
    cat("There are no GO being found.")
  }
  
  
  
  cat("### 4.2 KEGG pathway analysis \n\n")
  
  if(!(nrow(params$v$keggtab_up)==0)){
    cat("#### KEGG pathways enriched by upregular genes \n\n")
    cat("Here is a part of the result.\n\n")
    print(knitr::kable(head(params$v$keggtab_up)))
    cat("\n\n")
    
    par()
    a<-barplot(params$v$kegg_up)
    print(a)
    cat("\n\n")
    testupk<-FALSE
  }else{
    testupk<-TRUE
  }
  
  if(!(nrow(params$v$keggtab_down)==0)){
    cat("#### KEGG pathways enriched by Downregular genes \n\n")
    cat("Here is a part of the result.\n\n")
    print(knitr::kable(head(params$v$keggtab_down)))
    cat("\n\n")
    
    par()
    a<-barplot(params$v$kegg_down)
    print(a)
    cat("\n\n")
    testdownk<-FALSE
  }else{
    testdownk<-TRUE
  }
  
  
  if(testupk && testdownk){
    cat("There are no KEGG pathway being found. \n\n")
  }


}


```


```{r}
#part 1   

```
